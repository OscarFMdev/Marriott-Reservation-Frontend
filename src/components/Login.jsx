import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NavLink, useNavigate } from 'react-router-dom';
import { InputText } from 'primereact/inputtext';
import { Button } from 'primereact/button';
import { Password } from 'primereact/password';
import { signIn, getUser } from '../redux/reducer/Auth/auth';
import stl from './componentsCss/Login.module.css';

const Login = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { message } = useSelector((state) => state.auth);

  const handleSubmit = (e) => {
    e.preventDefault();
    const user = {
      email,
      password,
    };
    dispatch(signIn(user));
  };

  useEffect(() => {
    if (message === 'Logged in succesfully.') {
      dispatch(getUser());
      navigate('/rooms');
    }
  }, [message, navigate]);

  return (
    <div className="p-grid p-justify-center">
      <div className="p-col-12 p-md-8 p-lg-6">
        <div className={stl.card}>
          { !message ? null : <p className={stl.success}>{message}</p>}
          <h1>Login</h1>
          <form onSubmit={handleSubmit}>
            <div className="p-field pb-2">
              <span className="p-input-icon-right">
                <i className="pi pi-envelope" />
                <InputText
                  id="email"
                  type="email"
                  value={email}
                  onChange={(e) => setEmail(e.target.value)}
                  data-testid="email"
                  placeholder="Email"
                  required
                />
              </span>
            </div>

            <div className="p-field pb-3">

              <Password
                id="password"
                value={password}
                onChange={(e) => setPassword(e.target.value)}
                data-testid="password"
                placeholder="Password"
                required
                feedback={false}
                toggleMask
              />

            </div>
            <Button type="submit" label="Login" className="bg-green-500 border-200" />
            <p>
              Do not Have an Account?
              <span>
                <NavLink to="/signup"> Sign Up</NavLink>
              </span>
            </p>
          </form>
        </div>
      </div>
    </div>
  );
};

export default Login;
