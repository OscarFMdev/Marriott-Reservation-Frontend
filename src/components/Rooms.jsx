import { useSelector } from 'react-redux';
import { ProgressSpinner } from 'primereact/progressspinner';
import Carousel from './Carousel';
import stl from './componentsCss/Carousel.module.css';

const Rooms = () => {
  const rooms = useSelector((state) => state.rooms);

  if (rooms.status === 'loading') {
    return <ProgressSpinner className="progressSpinner" />;
  }
  return (
    <div className={stl.room_container}>
      <h2 className={stl.title}>Available Rooms</h2>
      {rooms.rooms.length === 0 && <h1>No Rooms</h1>}
      <Carousel />
    </div>
  );
};

export default Rooms;
