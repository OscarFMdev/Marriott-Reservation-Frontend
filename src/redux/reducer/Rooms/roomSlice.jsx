import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const BASEURL = 'https://marriott-reservation.onrender.com/api/v1';

const initialState = {
  rooms: [],
  status: '',
  message: '',
  error: '',
};

export const fetchRooms = createAsyncThunk('rooms/fetchRooms', async () => {
  const response = await fetch(`${BASEURL}/rooms`);
  const rooms = await response.json();
  return rooms;
});

export const addRoom = createAsyncThunk('rooms/addRoom', async (room) => {
  const response = await fetch(`${BASEURL}/rooms`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: localStorage.getItem('token'),
    },
    body: JSON.stringify({ room }),
  });

  const { status: code } = response;
  if (code === 201) {
    const { data, message } = await response.json();
    return {
      status: '',
      message,
      data,
    };
  }

  if (code === 401) {
    const { message } = await response.json();
    return {
      status: 'error',
      message,
    };
  }

  if (code === 422) {
    const { message } = await response.json();
    return {
      status: 'error',
      message,
    };
  }

  return null;
});
export const deleteRoom = createAsyncThunk('rooms/deleteRoom', async (id) => {
  const response = await fetch(`${BASEURL}/rooms/${id}`, {
    method: 'DELETE',
    headers: {
      Authorization: localStorage.getItem('token'),
    },
  });
  const deletedRoom = await response.json();
  return deletedRoom;
});

const roomsSlice = createSlice({
  name: 'rooms',
  initialState,
  reducers: {
    NewRoom: (state, action) => ({
      ...state,
      rooms: [...state.rooms, action.payload],
    }),
    PopRoom: (state, action) => ({
      ...state,
      rooms: state.rooms.filter((room) => room.id !== action.payload),
    }),
    setMessageEmpty: (state, action) => ({
      ...state,
      message: action.payload,
    }),
  },
  extraReducers: (builder) => {
    builder
      .addCase(addRoom.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(addRoom.fulfilled, (state, action) => ({
        ...state,
        rooms: [
          ...(action.payload.status === 201 ? action.payload.data : []),
          ...state.rooms,
        ],
        message: action.payload.message,
        status: action.payload.status,
      }))
      .addCase(addRoom.rejected, (state, action) => ({
        ...state,
        status: 'error',
        message: action.payload.message,
      }))
      .addCase(fetchRooms.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(fetchRooms.fulfilled, (state, action) => ({
        ...state,
        status: 'success',
        rooms: action.payload,
      }))
      .addCase(fetchRooms.rejected, (state, action) => ({
        ...state,
        status: 'error',
        message: action.payload.message,
      }))
      .addCase(deleteRoom.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(deleteRoom.fulfilled, (state, action) => ({
        ...state,
        status: 'success',
        rooms: state.rooms.filter((room) => room.id !== action.payload),
        message: action.payload.message,
      }))
      .addCase(deleteRoom.rejected, (state, action) => ({
        ...state,
        status: 'error',
        message: action.payload.message,
      }));
  },
});

export default roomsSlice.reducer;
export const { NewRoom, PopRoom, setMessageEmpty } = roomsSlice.actions;
