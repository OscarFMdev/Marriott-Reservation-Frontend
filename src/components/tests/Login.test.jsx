import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import Login from '../Login';
import Rooms from '../Rooms';
import store from '../../redux/configureStore';

describe('Login page integrity and operation', () => {
  test('Login title and button', () => {
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>,
    );
    expect(screen.queryAllByText(/Login/, { exact: true })).toHaveLength(2);
  });

  test('Login operation with wrong credentials', () => {
    render(
      <Provider store={store}>
        <Router>
          <Login />
          <Rooms />
        </Router>
      </Provider>,
    );
    const nameInput = screen.getByTestId('email');
    userEvent.type(nameInput, 'test@mail.com');
    const passwordInput = screen.getByTestId('password');
    userEvent.type(passwordInput, '123456');

    fireEvent.click(screen.getByRole('button', { name: 'Login' }));
    expect(screen.getByText('Available Rooms')).toBeDefined();
  });
});

test('Login renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Router>
        <Login />
      </Router>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
