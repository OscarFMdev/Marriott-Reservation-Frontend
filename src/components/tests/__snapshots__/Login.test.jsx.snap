// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login renders correctly 1`] = `
<div
  className="p-grid p-justify-center"
>
  <div
    className="p-col-12 p-md-8 p-lg-6"
  >
    <div
      className="_card_ac1e6f"
    >
      <p />
      <h1>
        Login
      </h1>
      <form
        onSubmit={[Function]}
      >
        <div
          className="p-field pb-2"
        >
          <span
            className="p-input-icon-right"
          >
            <i
              className="pi pi-envelope"
            />
            <input
              className="p-inputtext p-component"
              data-testid="email"
              id="email"
              onChange={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onPaste={[Function]}
              placeholder="Email"
              required={true}
              type="email"
              value=""
            />
          </span>
        </div>
        <div
          className="p-field pb-3"
        >
          <div
            className="p-password p-component p-inputwrapper p-input-icon-right"
            id="password"
            style={null}
          >
            <input
              className="p-inputtext p-component p-password-input"
              data-testid="password"
              id={null}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
              placeholder="Password"
              required={true}
              style={null}
              type="password"
              value=""
            />
            <i
              className="pi pi-eye"
              onClick={[Function]}
            />
          </div>
        </div>
        <button
          aria-label="Login"
          className="p-button p-component bg-green-500 border-200"
          disabled={false}
          type="submit"
        >
          <span
            className="p-button-label p-c"
          >
            Login
          </span>
        </button>
        <p>
          Do not Have an Account?
          <span>
            <a
              className=""
              href="/signup"
              onClick={[Function]}
            >
               Sign Up
            </a>
          </span>
        </p>
      </form>
    </div>
  </div>
</div>
`;
